src/roman_datamodels/stnode/core/_d_node.py

```diff
@@ -0,0 +1,424 @@
+        variable names.
+        """
+        # Private values should have already been handled by the __getattribute__ method
+        #   bail out if we are falling back on this method
+        if key.startswith("_"):
+            raise AttributeError(f"No attribute {key}")
+
+        return self._pull_node(key)
+
+    def __getitem__(self, key: str) -> _T:
+        """Dictionary style access data"""
+        # Try to directly access the attribute
+        try:
+            return self._pull_node(key)
+
+        except MissingFieldError as err:
+            # For lazy nodes, we need to check if the key is in the fields and
+            # grab the value from the property
+            if (key_ := self._to_field_key(key)) in self.fields:
+                value: _T = getattr(self, key_)
+                return value
+
+            raise KeyError(f"No such attribute ({key}) found in node") from err
+
+    def _check_key(self, key: str) -> bool:
+        """
+        Check if the key is settable
+        """
+        return self._to_schema_key(key) in self._data or self._to_field_key(key) in self.fields
+
+    def _set_node(self, key: str, value: _T, check: bool = True, wrap: bool = True) -> None:
+        """
+        Attempt to set a value in for the node, wrapping data if necessary.
+        """
+        # Private keys should just be in the normal __dict__
+        if key[0] == "_":
+            self.__dict__[key] = value
+        else:
+            if check and not self._check_key(key):
+                raise AttributeError(f"No such attribute ({key}) found in node")
+
+            self._data[self._to_schema_key(key)] = self._wrap_into_node(key, value, wrap=wrap)
+
+    def __setattr__(self, key: str, value: _T) -> None:
+        """
+        Permit assigning dict keys as attributes.
+        """
+        self._set_node(key, value)
+
+    def __setitem__(self, key: str, value: _T) -> None:
+        """Dictionary style access set data"""
+        self._set_node(key, value, check=False, wrap=False)
+
+    def _get_node(self, key: str, default: Callable[[], _T]) -> _T:
+        """
+        Get a node and if not found construct it with the default value.
+
+        Note: This is what creates the lazy instantiation of the default values
+        This is what all the schema related properties should use to get their values.
+
+        Parameters
+        ----------
+        key : str
+            The key to look for in the dictionary.
+        default : Callable
+            The constructor for a default. Its callable to allow for lazy instantiation.
+            This will often take the form of a lambda, that way the argument to the lambda
+            is not evaluated until the default is actually needed, saving time and memory.
+            for things like numpy arrays.
+        """
+
+        if not self._has_node(key):
+            self._set_node(key, default())
+
+        return self._pull_node(key)
```

Comment:
This method is the backbone addition to DNode that enables the lazy magic. It allows it a value to be set using a "default" function provided it does not exist.